File: santa-tracker/src/main/java/com/google/android/apps/santatracker/launch/adapters/CardAdapter.java
/*
 * Copyright 2019. Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.santatracker.launch.adapters;

import android.content.res.Resources;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.ColorRes;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.core.content.ContextCompat;
import androidx.core.view.ViewCompat;
import androidx.recyclerview.widget.RecyclerView;
import com.google.android.apps.santatracker.R;
import com.google.android.apps.santatracker.launch.AbstractLaunch;
import com.google.android.apps.santatracker.launch.CardLayoutManager;
import com.google.android.apps.santatracker.launch.LaunchCollection;
import com.google.android.apps.santatracker.launch.LauncherDataChangedCallback;
import com.google.android.apps.santatracker.launch.SantaContext;
import com.google.android.flexbox.FlexboxLayoutManager;

public class CardAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder>
        implements LauncherDataChangedCallback {

    private static final int HEADER_VIEW_TYPE = 0;
    private static final int ITEM_VIEW_TYPE = 1;

    private static final int[] LOCKED_COLORS =
            new int[] {
                R.color.SantaTranslucentPurple,
                R.color.SantaTranslucentBlue,
                R.color.SantaTranslucentYellow,
                R.color.SantaTranslucentGreen,
                R.color.SantaTranslucentRed,
            };

    private SantaContext mContext;
    @Nullable LaunchCollection mLaunchers;
    @StringRes private int mHeaderTextRes;

    public CardAdapter(SantaContext santaContext, @StringRes int headerTextRes) {
        mContext = santaContext;
        mHeaderTextRes = headerTextRes;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder holder;
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View itemView;
        switch (viewType) {
            case HEADER_VIEW_TYPE:
                itemView = inflater.inflate(R.layout.layout_village_header, parent, false);
                holder =
                        new HeaderViewHolder(
                                itemView, mContext.getResources().getString(mHeaderTextRes));
                break;
            case ITEM_VIEW_TYPE:
                itemView = inflater.inflate(R.layout.layout_village_card, parent, false);
                holder = new ItemViewHolder(itemView);
                break;
            default:
                throw new IllegalArgumentException();
        }
        return holder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        int viewType = getItemViewType(position);
        switch (viewType) {
            case HEADER_VIEW_TYPE:
                break;
            case ITEM_VIEW_TYPE:
                bindItemViewHolder((ItemViewHolder) holder, position);
                break;
            default:
                throw new IllegalArgumentException();
        }
    }

    public void setLaunchers(LaunchCollection launchers) {
        mLaunchers = launchers;
        mLaunchers.updateVisibleList();
    }

    void bindItemViewHolder(ItemViewHolder holder, int position) {
        int lockedLayerColor =
                ContextCompat.getColor(
                        holder.itemView.getContext(), getLockedViewResource(position));
        AbstractLaunch launcher = mLaunchers.getVisibleLauncherFromPosition(position);
        holder.setLauncher(launcher, lockedLayerColor, launcher.isLocked());
        boolean isFeatured = isCardFeatured(position);
        ViewGroup.LayoutParams lp = holder.itemView.getLayoutParams();
        if (lp instanceof FlexboxLayoutManager.LayoutParams) {
            FlexboxLayoutManager.LayoutParams flexboxLp = (FlexboxLayoutManager.LayoutParams) lp;
            Resources resources = mContext.getResources();
            if (isFeatured) {
                flexboxLp.setFlexBasisPercent(0.4f);
                int minWidth =
                        resources.getDimensionPixelSize(R.dimen.village_card_featured_min_width);
                int space = resources.getDimensionPixelSize(R.dimen.spacing_extra_large);
                flexboxLp.setMinWidth(minWidth);
                holder.itemView.setPadding(space, 0, space, space);
            } else {
                flexboxLp.setFlexBasisPercent(0.24f);
                int minWidth =
                        resources.getDimensionPixelSize(
                                R.dimen.village_card_non_featured_min_width);
                int maxWidth = resources.getDimensionPixelSize(R.dimen.village_card_max_width);
                flexboxLp.setMinWidth(minWidth);
                flexboxLp.setMaxWidth(maxWidth);
                int space = resources.getDimensionPixelSize(R.dimen.spacing_large);
                holder.itemView.setPadding(space, 0, space, space);
            }
            flexboxLp.setFlexGrow(1f);
            holder.itemView.setLayoutParams(flexboxLp);
        }
        holder.launcher.setLockedView(holder.lockedView);
        holder.launcher.applyState();
        holder.launcher.setImageView(holder.backgroundImageView);
        ViewCompat.setTransitionName(holder.backgroundImageView, holder.launcher.getTitle());
    }

    @Override
    public int getItemCount() {
        if (mLaunchers != null) {
            return mLaunchers.getNumVisibleLaunchers();
        }
        return 0;
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0) {
            return HEADER_VIEW_TYPE;
        } else {
            return ITEM_VIEW_TYPE;
        }
    }

    @Override
    public void refreshData() {
        if (mLaunchers != null) {
            mLaunchers.updateVisibleList();
        }
        notifyDataSetChanged();
    }

    public boolean isCardFeatured(int position) {
        return mLaunchers.getVisibleLauncherFromPosition(position).isFeatured();
    }

    @ColorRes
    private int getLockedViewResource(int position) {
        int lockedColor =
                mLaunchers.getVisibleLauncherFromPosition(position).getLockedViewResource();
        if (lockedColor != -1) {
            return lockedColor;
        } else {
            return LOCKED_COLORS[position % LOCKED_COLORS.length];
        }
    }

    public static class ColumnCountCalculator {
        private int nonFeatureColumnSize;
        private int featureColumnSize;
        private LaunchCollection launches;

        public ColumnCountCalculator(
                int nonFeatureColumnSize, int featureColumnSize, LaunchCollection launches) {
            this.launches = launches;
            this.nonFeatureColumnSize = nonFeatureColumnSize;
            this.featureColumnSize = featureColumnSize;
        }

        public int calculateColumnCount(int position) {
            return CardLayoutManager.SPAN_COUNT
                    / (launches.getVisibleLauncherFromPosition(position).isFeatured()
                            ? featureColumnSize
                            : nonFeatureColumnSize);
        }
    }
}

File: tracker/src/main/java/com/google/android/apps/santatracker/tracker/ui/CardAdapter.kt
/*
 * Copyright 2019. Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.santatracker.tracker.ui

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.google.android.apps.santatracker.tracker.R
import com.google.android.apps.santatracker.tracker.time.Clock
import com.google.android.apps.santatracker.tracker.vo.Destination
import com.google.android.apps.santatracker.tracker.vo.DestinationStreetView
import com.google.android.apps.santatracker.tracker.vo.StreamEntry
import com.google.android.apps.santatracker.tracker.vo.TrackerCard

class CardAdapter(
    val clock: Clock,
    private val callback: Callback,
    recyclerView: RecyclerView,
    var disableDestinationPhoto: Boolean,
    private val isTv: Boolean
) : RecyclerView.Adapter<CardViewHolder>() {

    interface Callback {
        fun onPlayMovie(youtubeId: String)
        fun onStreetView(streetView: DestinationStreetView)
    }

    companion object {
        const val TYPE_DASHBOARD = 1
        const val TYPE_DESTINATION = 2
        const val TYPE_FACTOID = 3
        const val TYPE_MOVIE = 4
        const val TYPE_PHOTO = 5
        const val TYPE_UPDATE = 6
    }

    private var cards = emptyList<TrackerCard>()

    val dashboard: DashboardViewHolder =
            DashboardViewHolder(LayoutInflater.from(recyclerView.context), recyclerView)

    private val onClickListener = View.OnClickListener { view ->
        when (view.id) {
            R.id.movie_play -> callback.onPlayMovie(view.tag as String)
            R.id.destination_street_view -> callback.onStreetView(view.tag as DestinationStreetView)
        }
    }

    init {
        setHasStableIds(true)
    }

    override fun getItemCount() = 1 + cards.size

    override fun getItemId(position: Int) = when (position) {
        0 -> -1
        else -> cardAt(position).value
    }

    override fun getItemViewType(position: Int): Int {
        if (position == 0) {
            return TYPE_DASHBOARD
        }
        val card = cardAt(position)
        if (card is Destination) {
            return TYPE_DESTINATION
        }
        val streamEntry = card as StreamEntry
        return when (streamEntry.type) {
            StreamEntry.TYPE_DID_YOU_KNOW -> TYPE_FACTOID
            StreamEntry.TYPE_YOUTUBE_ID -> TYPE_MOVIE
            StreamEntry.TYPE_IMAGE_URL -> TYPE_PHOTO
            StreamEntry.TYPE_STATUS -> TYPE_UPDATE
            else -> -1
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CardViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return when (viewType) {
            TYPE_DASHBOARD -> dashboard
            TYPE_DESTINATION -> DestinationViewHolder(
                    inflater, parent, onClickListener, clock, disableDestinationPhoto, !isTv)
            TYPE_FACTOID -> FactoidViewHolder(inflater, parent)
            TYPE_MOVIE -> MovieViewHolder(inflater, parent, onClickListener, !isTv)
            TYPE_PHOTO -> PhotoViewHolder(inflater, parent)
            TYPE_UPDATE -> UpdateViewHolder(inflater, parent)
            else -> throw IllegalArgumentException()
        }
    }

    override fun onBindViewHolder(holder: CardViewHolder, position: Int) {
        if (position == 0) {
            return
        }
        holder.bind(cardAt(position))
    }

    fun setCards(cards: List<TrackerCard>) {
        this.cards = cards
        notifyDataSetChanged()
    }

    private fun cardAt(position: Int) = cards[position - 1]
}
